src/parsing/syntax/ast_lint.mbt

```diff
@@ -0,0 +1,245 @@
+    tys=@list.empty(),
+    loc=no_location,
+  )
+  let params = match params {
+    None => Some(@list.empty())
+    Some(_) => params // no check!
+  }
+  TopFuncDef(
+    fun_decl=FunDecl::{
+      type_name: None,
+      has_error: None,
+      is_async: false,
+      name: Binder::{ name: test_name, loc: no_location },
+      decl_params: params,
+      params_loc: no_location,
+      quantifiers: @list.empty(),
+      return_type: Some(unit_type),
+      error_type: ErrorType::ErrorType(ty=string_type),
+      vis: Default,
+      attrs: @list.empty(),
+      doc: DocString::empty(),
+    },
+    decl_body=DeclBody(expr~, local_types~),
+    loc~,
+  )
+}
+
+///|
+pub fn post_process(impls : Impls) -> Impls {
+  let env = AstLintEnv(())
+  fn go(i : Int, impls : Impls) -> Impls {
+    match impls {
+      Empty => @list.empty()
+      More(impl_, tail=rest) =>
+        match env.visit_impl(impl_) {
+          TopTest(expr~, params~, local_types~, loc~, ..) => {
+            let impl_ = test_to_func(i, expr, params, local_types, loc)
+            go(i + 1, rest).add(impl_)
+          }
+          TopFuncDef(
+            fun_decl={ name: { name: "init", .. }, .. },
+            decl_body=DeclBody(expr~, local_types~),
+            loc~
+          ) => {
+            let impl_ = TopExpr(expr~, is_main=false, local_types~, loc~)
+            go(i, rest).add(impl_)
+          }
+          TopFuncDef(
+            fun_decl={ name: { name: "main", .. }, type_name: None, attrs, .. },
+            decl_body=DeclBody(expr~, local_types~),
+            loc~
+          ) => {
+            let is_test_main_attribute = fn(attr : Attribute) {
+              match attr {
+                { parsed: Some(Ident({ qual: None, name: "test_entry" })), .. } =>
+                  true
+                { parsed: None, .. } => false // TODO: replace with panic() after complete attribute_parser
+                _ => false
+              }
+            }
+            if @basic.test_mode.val && !attrs.any(is_test_main_attribute) {
+              go(i, rest)
+            } else {
+              go(i, rest).add(TopExpr(expr~, is_main=true, local_types~, loc~))
+            }
+          }
+          TopTypeDef(decl) as top => {
+            let external = decl.attrs.find(fn(attr : Attribute) {
+              match attr.parsed {
+                Some(Ident({ qual: None, name: "external" })) => true
+                _ => false
+              }
+            })
+            if external is Some(_) {
+              let decl = match decl.components {
+                Abstract | Extern => { ..decl, components: Extern }
+                _ => decl
+              }
+              go(i, rest).add(TopTypeDef(decl))
+            } else {
+              go(i, rest).add(top)
+            }
+          }
+          _ => go(i, rest).add(impl_)
+        }
+    }
+  }
+
+  let impls = go(0, impls)
+  fn filter_impls_by_cfg(impls : Impls) -> Impls {
```

Comment:
`filter_impls_by_cfg` is currently unnecessary. We can ignore the cfg attribute.